File: C-Plus-Plus/dp/longest_common_subsequence.cpp
/*
Given two sequences, find the length of longest subsequence present in both of them.
A subsequence is a sequence that appears in the same relative order, but not necessarily contiguous.
*/

#include <iostream>
#include <vector> // for 2D vector
using namespace std;

int lowest_common_subsequence(string &x, string &y)
{
    int len_x = x.length();
    int len_y = y.length();

    vector<vector<int> > dp (len_x+1, vector<int> (len_y+1,0));

    for(int i=1;i<=len_x;i++)
    {
        for(int j=1;j<=len_y;j++)
        {
            int q=0;

            if(x[i-1]==y[j-1]) //if matches
                q=1+dp[i-1][j-1];
            else     //if doesn't match
                q=max(dp[i-1][j],dp[i][j-1]);

        dp[i][j]=q;
        }
    }

    return dp[len_x][len_y]; //return last element
}

int main()
{
    string str1,str2;
    cout<<"Enter string1 and string2: ";
    cin>>str1>>str2;
    cout<<"output : "<<lowest_common_subsequence(str1,str2)<<endl;
}

/*
   Enter string1 and string2 : ABCDEFGH BDHCEG
   output : 4
*/

// Time and Space complexity : O(mn), where m = length of string1 and n = length of string2

File: Go/dp/longest_common_subsequence/longest_common_subsequence.go
/* Given two sequences, find the length of longest subsequence present
in both of them. A subsequence is a sequence that appears in the
same relative order, but not necessarily contiguous.
For example, “abc”, “abg”, “bdf”, “aeg”, ‘”acefg”, .. etc
are subsequences of “abcdefg”. So, the answer will be the longest
length i.e. 5
*/

package main

import "fmt"

// function to find the maximum
func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

// longest common subsequence function
func lcs(X, Y string, m, n int) int {
	// creating 2d array
	L := make([][]int, m+1)
	for itr := 0; itr <= m; itr++ {
		L[itr] = make([]int, n+1)
	}

	// bottom up fashion
	for i := 0; i <= m; i++ {
		for j := 0; j <= n; j++ {
			if i == 0 || j == 0 {
				L[i][j] = 0
			} else if X[i-1] == Y[j-1] {
				L[i][j] = L[i-1][j-1] + 1
			} else {
				L[i][j] = max(L[i-1][j], L[i][j-1])
			}
		}
	}

	return L[m][n]
}

func main() {
	var str1 string
	var str2 string

	fmt.Print("Enter the 1st string : ")
	fmt.Scan(&str1)

	fmt.Print("Enter the 2nd string : ")
	fmt.Scan(&str2)

	m := len(str1)
	n := len(str2)

	if m < n {
		fmt.Println("Length of LCS is: ", lcs(str1, str2, m, n))
	} else {
		fmt.Println("Length of LCS is: ", lcs(str2, str1, n, m))
	}
}

/*
TIME COMPLEXITY: O(M*N)
SPACE COMPLEXITY: O(M*N)


OUTPUT

TEST CASE 1:
Enter the 1st string : NEOALGO
Enter the 2nd string : ALGORITHM
Length of LCS is:  4

TEST CASE 2:
Enter the 1st string : ABQWERTYCD
Enter the 2nd string : QWERTY
Length of LCS is:  6
*/
